File: api/like.py
"""
api/like.py  —  Vercel serverless function
Handles POST (like/unlike) requests from the browser.
The GitHub PAT lives only in Vercel's environment variables.
Never touches the client.
"""

import json
import os
import base64
from http.server import BaseHTTPRequestHandler


GITHUB_TOKEN = os.environ.get("GITHUB_TOKEN", "")
GITHUB_REPO  = os.environ.get("GITHUB_REPO", "")   # e.g. "ailtotanvir/curiosity-digest"
LIKES_PATH   = "likes.json"
BRANCH       = "main"
API_BASE     = f"https://api.github.com/repos/{GITHUB_REPO}/contents/{LIKES_PATH}"

HEADERS = {
    "Authorization": f"token {GITHUB_TOKEN}",
    "Accept":        "application/vnd.github+json",
    "X-GitHub-Api-Version": "2022-11-28",
    "Content-Type":  "application/json",
}


def gh_get():
    """Fetch current likes.json from GitHub. Returns (data_dict, sha)."""
    import urllib.request
    req = urllib.request.Request(API_BASE, headers=HEADERS)
    try:
        with urllib.request.urlopen(req) as r:
            body = json.loads(r.read())
            content = json.loads(base64.b64decode(body["content"]).decode())
            return content, body["sha"]
    except Exception as e:
        # File doesn't exist yet
        return {}, None


def gh_put(likes: dict, sha: str | None, action: str, post_id: str):
    """Write updated likes.json back to GitHub."""
    import urllib.request
    encoded = base64.b64encode(
        json.dumps(likes, indent=2, ensure_ascii=False).encode()
    ).decode()

    payload = {
        "message": f"likes: {action} {post_id}",
        "content": encoded,
        "branch":  BRANCH,
    }
    if sha:
        payload["sha"] = sha

    data = json.dumps(payload).encode()
    req  = urllib.request.Request(API_BASE, data=data, headers=HEADERS, method="PUT")
    with urllib.request.urlopen(req) as r:
        return json.loads(r.read())


class handler(BaseHTTPRequestHandler):

    def do_OPTIONS(self):
        """CORS preflight."""
        self.send_response(200)
        self._cors()
        self.end_headers()

    def do_POST(self):
        length = int(self.headers.get("Content-Length", 0))
        body   = json.loads(self.rfile.read(length))

        action  = body.get("action", "like")   # "like" or "unlike"
        post_id = body.get("id", "")

        if not post_id:
            self._respond(400, {"error": "missing id"})
            return

        if not GITHUB_TOKEN or not GITHUB_REPO:
            self._respond(500, {"error": "server not configured"})
            return

        try:
            likes, sha = gh_get()

            if action == "unlike":
                likes.pop(post_id, None)
            else:
                from datetime import datetime, timezone
                likes[post_id] = {
                    "id":      post_id,
                    "title":   body.get("title", ""),
                    "source":  body.get("source", ""),
                    "date":    body.get("date", ""),
                    "cat":     body.get("cat", ""),
                    "link":    body.get("link", ""),
                    "summary": body.get("summary", ""),
                    "likedAt": datetime.now(timezone.utc).isoformat(),
                }

            gh_put(likes, sha, action, post_id)
            self._respond(200, {"ok": True, "total": len(likes)})

        except Exception as e:
            self._respond(500, {"error": str(e)})

    def _cors(self):
        self.send_header("Access-Control-Allow-Origin",  "*")
        self.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")

    def _respond(self, status: int, data: dict):
        body = json.dumps(data).encode()
        self.send_response(status)
        self._cors()
        self.send_header("Content-Type",   "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def log_message(self, *_):
        pass   # suppress default access log noise

File: api/like.js
// api/like.js — Vercel serverless function (Node.js)
// Handles POST /api/like from the browser.
// GITHUB_TOKEN and GITHUB_REPO live in Vercel env vars only — never in HTML.

const https = require("https");

const GH_TOKEN = process.env.GITHUB_TOKEN || "";
const GH_REPO  = process.env.GITHUB_REPO  || "";
const FILE     = "likes.json";
const BRANCH   = "main";

// Explicitly build the path — no URL parsing, no missing protocol
const API_HOSTNAME = "api.github.com";
const API_PATH     = `/repos/${GH_REPO}/contents/${FILE}`;

const GH_HEADERS = {
  Authorization:          `token ${GH_TOKEN}`,
  Accept:                 "application/vnd.github+json",
  "X-GitHub-Api-Version": "2022-11-28",
  "User-Agent":           "curiosity-digest-likes",
  "Content-Type":         "application/json",
};

// ── Tiny promise wrapper around https ────────────────────────────────────────

function request(method, extraHeaders, body) {
  return new Promise((resolve, reject) => {
    const options = {
      hostname: API_HOSTNAME,
      path:     API_PATH,
      method,
      headers:  { ...GH_HEADERS, ...extraHeaders },
    };
    const req = https.request(options, (res) => {
      let data = "";
      res.on("data", (c) => (data += c));
      res.on("end", () => {
        try { resolve({ status: res.statusCode, body: JSON.parse(data) }); }
        catch { resolve({ status: res.statusCode, body: data }); }
      });
    });
    req.on("error", reject);
    if (body) req.write(body);
    req.end();
  });
}

// ── Read current likes.json ───────────────────────────────────────────────────

async function getLikes() {
  const res = await request("GET", {});
  if (res.status === 404) return { likes: {}, sha: null };
  if (res.status !== 200) throw new Error(`GitHub GET ${res.status}: ${JSON.stringify(res.body)}`);
  const likes = JSON.parse(Buffer.from(res.body.content, "base64").toString());
  return { likes, sha: res.body.sha };
}

// ── Write updated likes.json ──────────────────────────────────────────────────

async function putLikes(likes, sha, action, id) {
  const content = Buffer.from(JSON.stringify(likes, null, 2)).toString("base64");
  const payload = JSON.stringify({
    message: `likes: ${action} ${id}`,
    content,
    branch: BRANCH,
    ...(sha ? { sha } : {}),
  });
  const res = await request("PUT", { "Content-Length": Buffer.byteLength(payload) }, payload);
  if (res.status !== 200 && res.status !== 201)
    throw new Error(`GitHub PUT ${res.status}: ${JSON.stringify(res.body)}`);
  return res.body;
}

// ── Handler ───────────────────────────────────────────────────────────────────

module.exports = async function handler(req, res) {
  res.setHeader("Access-Control-Allow-Origin",  "*");
  res.setHeader("Access-Control-Allow-Methods", "POST, OPTIONS");
  res.setHeader("Access-Control-Allow-Headers", "Content-Type");

  if (req.method === "OPTIONS") return res.status(200).end();
  if (req.method !== "POST")   return res.status(405).json({ error: "POST only" });

  if (!GH_TOKEN || !GH_REPO)
    return res.status(500).json({ error: "Server not configured — check Vercel env vars" });

  const body   = req.body || {};
  const action = body.action || "like";
  const id     = body.id || "";

  if (!id) return res.status(400).json({ error: "missing id" });

  try {
    const { likes, sha } = await getLikes();

    if (action === "unlike") {
      delete likes[id];
    } else {
      likes[id] = {
        id,
        title:   body.title   || "",
        source:  body.source  || "",
        date:    body.date    || "",
        cat:     body.cat     || "",
        link:    body.link    || "",
        summary: body.summary || "",
        likedAt: new Date().toISOString(),
      };
    }

    await putLikes(likes, sha, action, id);
    return res.status(200).json({ ok: true, total: Object.keys(likes).length });

  } catch (err) {
    console.error(err);
    return res.status(500).json({ error: err.message });
  }
};
